design/src/rfcs/rfc0039_behavior_major_versions.md

```diff
@@ -0,0 +1,89 @@
+<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
+RFC: Behavior Major Versions
+=============
+
+<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
+> Status: RFC
+>
+> Applies to: client
+
+<!-- A great RFC will include a list of changes at the bottom so that the implementor can be sure they haven't missed anything -->
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+<!-- Insert a short paragraph explaining, at a high level, what this RFC is for -->
+This RFC describes "Behavior Major Versions," a mechanism to allow SDKs to ship breaking behavioral changes like a new retry strategy, while allowing customers who rely on extremely consistent behavior to evolve at their own pace.
+
+By adding behavior major versions (BMV) to the Rust SDK, we will make it possible to ship new secure/recommended defaults to new customers without impacting legacy customers.
+
+The fundamental issue stems around our inability to communicate and decouple releases of service updates and behavior within a single major version.
+
+Both legacy and new SDKs have the need to alter their SDKs default. Historically, this caused new customers on legacy SDKs to be subject to legacy defaults, even when a better alternative existed.
+
+For new SDKs, a GA cutline presents difficult choices around timeline and features that can’t be added later without altering behavior.
+
+Both of these use cases are addressed by Behavior Major Versions.
+
+<!-- The "Terminology" section is optional but is really useful for defining the technical terms you're using in the RFC -->
+Terminology
+-----------
+
+- **Some Term**: A definition for that term
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+In the current version of the SDK, users can construct clients without indicating any sort of behavior major version.
+Once this RFC is implemented, there will be three ways to set a behavior major version:
+
+1. In code via `aws_config::from_env_with_version(BehaviorMajorVersion::latest())` and `<service>::Config::builder().behavior_major_version(...)`
+2. By enabling `behavior-version-latest` in either `aws-config` (which brings back `from_env`) OR a specific generated SDK crate
+
+```toml
+# Cargo.toml
+[dependencies]
+aws-config = { version = "1", features = ["behavior-version-latest"] }
+# OR
+aws-sdk-s3 = { version = "1", features = ["behavior-version-latest"] }
+```
+
+If no `BehaviorMajorVersion` is set, the client will panic during construction.
+
+`BehaviorMajorVersion` is an opaque struct with initializers like `::latest()`, `::v2023_11_09()`. Downstream code can check the version by calling methods like `::supports_v1()`
```

Comment:
I don't think the `supports_*()` would work or be useful since the function wouldn't exist in a sufficiently old SDK. I think if we were to support something like this, we would need to have `from_str` and `supports_str` methods on the BMV.

Reply:
sufficently old SDKs won't have the newer code so there's no new versions to suppress